lib/exhal/form.ex

```diff
@@ -0,0 +1,140 @@
+defmodule ExHal.Form do
+  @moduledoc """
+
+  Represents a [Dwolla style HAL
+  form](https://github.com/Dwolla/hal-forms). Generally these are
+  acquired from `ExHal.Document.get_form/2`,
+  `ExHal.Document.fetch_form/2`, etc
+
+  """
+
+  alias ExHal.{
+    FormField,
+    JsonFormEncoder
+  }
+
+  @typedoc """
+  A from that can be completed and submitted. This type is opaque and
+  should use only as a argument to functions in this `ExHal.Form` module.
+  """
+  @opaque t :: %__MODULE__{}
+
+  defstruct [
+    :target,
+    :method,
+    :content_type,
+    :fields
+  ]
+
+  @spec from_forms_entry(%{}) :: __MODULE__.t()
+  @doc """
+  Creates a new form from raw form JSON.
+
+  Raises `ArgumentError` if `a_map` is not a valid form fragment.
+  """
+  def from_forms_entry(a_map) do
+    %__MODULE__{
+      target: extract_target(a_map),
+      method: extract_method(a_map),
+      content_type: extract_content_type(a_map),
+      fields: extract_fields(a_map)
+    }
+  end
+
+  @spec get_fields(__MODULE__.t()) :: [FormField.t()]
+  @doc """
+  Returns list of the fields in this form.
+  """
+  def get_fields(a_form) do
+    a_form.fields
+  end
+
+  @spec set_field_value(__MODULE__.t(), String.t(), FormField.field_value()) :: __MODULE__.t()
+  @doc """
+  Returns form with the specified fields value updated.
+
+  Raises `ArgumentError` if the specified field doesn't exist.
+  """
+  def set_field_value(form, field_name, new_value) do
+    updated_field =
+      get_field(form, field_name)
+      |> FormField.set_value(new_value)
+
+    replace_field(form, field_name, updated_field)
+  end
+
+  @spec submit(__MODULE__.t(), Client.t()) ::
+          {:ok, Document.t(), ResponseHeader.t()} | {:error, ExHal.Error.t()}
```

Comment:
I believe this may also return `{:error, Document.t(), ResponseHeader.t()}`

Reply:
good catch. fixed.